File: ObjCDemo/platforms/ios/CordovaDemo/Plugins/HelloPlugin.m
//
//  HelloPlugin.m
//  CordovaDemo
//
//  Created by Fay on 2017/4/13.
//
//

#import "HelloPlugin.h"
#import "ExtraViewController.h"
#import "AppDelegate.h"

@interface HelloPlugin () <ExtraViewControllerDelegate>

@end

@implementation HelloPlugin

#pragma mark - Cordova Plugin Methods

//JS -> OC: ObjC callback method
- (void)sayHello:(CDVInvokedUrlCommand *)command
{
    CDVPluginResult* pluginResult = nil;
    if(!command.arguments || command.arguments.count <= 0) {
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR];
    } else {
        NSLog(@"%@",command.arguments );
        
        //CDVCommandStatus_OK: js_success_callback
        //CDVCommandStatus_ERROR: js_failure_callback
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:@"Hello, I am ObjC!"];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }
}

//JS -> OC: ObjC callback method
- (void)openNew:(CDVInvokedUrlCommand *)command
{
    ExtraViewController *evc = [[ExtraViewController alloc] init];
    evc.delegate = self;
    AppDelegate *appDelegate = (AppDelegate *)[UIApplication sharedApplication].delegate;
    [appDelegate.viewController presentViewController:evc animated:YES completion:NULL];
}


#pragma mark - ExtraViewControllerDelegate Methods

- (void)extraViewControllerDidDismiss:(ExtraViewController *)extraViewController
{
    //OC -> JS
    [self.commandDelegate evalJs:[NSString stringWithFormat:@"app.dismissViewController('View controller has been dismiss.');"]];
}

@end

File: Objective-C & Swift/SwiftDemo/platforms/ios/CordovaDemo/Plugins/HelloPlugin.swift
//
//  HelloPlugin.swift
//  CordovaDemo
//
//  Created by Fay on 2017/4/13.
//
//

import Foundation

@objc(HelloPlugin) class HelloPlugin : CDVPlugin {
    
    //Swift callback method
    @objc(sayHello:)
    func sayHello(command: CDVInvokedUrlCommand) {
        var pluginResult: CDVPluginResult
        if command.arguments == nil || command.arguments.count <= 0 {
            pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR)
        } else {
            print(command.arguments)
            
            //CDVCommandStatus_OK: js_success_callback
            //CDVCommandStatus_ERROR: js_failure_callback
            pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs:"Hello, I am Swift!")
            self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
        }
    }
}
